lib/functoria/impl.ml

```diff
@@ -53,84 +60,376 @@ let main ?packages ?packages_v ?keys ?extra_deps module_name ty =
+
+type ex = Ex : 'a -> ex
 
-let hash_abstract (Abstract x) = hash x
+let equal_list p l1 l2 =
+  List.length l1 = List.length l2 && List.for_all2 p l1 l2
 
-let rec equal : type t1 t2. t1 t -> t2 t -> bool =
+let rec equal : type t1 t2. t1 t -> t2 t -> (t1, t2) Eq.eq =
  fun x y ->
+  let ( &&! ) a b = Eq.andb b a in
   match (x, y) with
-  | Dev c, Dev c' -> Device.equal c c'
-  | App a, App b -> equal a.f b.f && equal a.x b.x
-  | If (cond1, t1, e1), If (cond2, t2, e2) ->
-      (* Key.value is a functional value (it contains a closure for eval).
-         There is no prettier way than physical equality. *)
-      cond1 == cond2 && equal t1 t2 && equal e1 e2
-  | _ -> false
+  | Dev c, Dev c' -> (
+      let b =
+        equal_tl c.args c'.args (Device.witness c.dev c'.dev)
+        &&! equal_list equal_abstract c.deps c'.deps
+      in
+      match b with Eq -> Eq | NotEq -> NotEq )
+  | App a, App b -> (
+      let b = equal_tl a.args b.args (equal a.f b.f) in
+      match b with Eq -> Eq | NotEq -> NotEq )
+  | If x1, If x2 -> (
+      let b =
+        (* Key.value is a functional value (it contains a closure for eval).
+           There is no prettier way than physical equality. *)
+        equal x1.default x2.default
+        &&! (Obj.repr x1.cond == Obj.repr x2.cond)
+        &&! equal_list
+              (fun (p1, t1) (p2, t2) ->
+                Ex p1 = Ex p2 && equal_abstract (abstract t1) (abstract t2))
+              x1.branches x2.branches
+      in
+      match b with Eq -> Eq | NotEq -> NotEq )
+  | _ -> NotEq
+
+and equal_abstract (Abstract x) (Abstract y) = Eq.to_bool @@ equal x y
 
-let equal_abtract (Abstract a) (Abstract b) = equal a b
+and equal_tl :
+    type t1 t2 v1 v2.
+    (t1, v1) tl -> (t2, v2) tl -> (t1, t2) Eq.eq -> (v1, v2) Eq.eq =
+ fun x y eq ->
+  match (x, y, eq) with
+  | Nil, Nil, Eq -> Eq
+  | Cons (h1, t1), Cons (h2, t2), Eq -> (
+      let b = Eq.andeq (equal h1 h2) (equal_tl t1 t2 Eq) in
+      match b with Eq -> Eq | NotEq -> NotEq )
+  | _ -> NotEq
 
 module Tbl = Hashtbl.Make (struct
   type t = abstract
 
   let hash = hash_abstract
 
-  let equal = equal_abtract
+  let equal = equal_abstract
 end)
 
-type 'b f_dev = { f : 'a. ('a, abstract) Device.t -> 'b }
+module HC : sig
+  type tbl
+
+  val create : unit -> tbl
+
+  val add : tbl -> 'a t -> 'a t -> unit
+
+  val get : tbl -> 'a t -> 'a t option
+end = struct
+  type tbl = abstract Tbl.t
 
-type 'a f_if = cond:bool Key.value -> then_:'a -> else_:'a -> 'a
+  let create () = Tbl.create 50
 
-type 'a f_app = f:'a -> x:'a -> 'a
+  let add tbl a b = Tbl.add tbl (abstract a) (abstract b)
 
-let with_left_most_device ctx t { f } =
+  let get (type a) tbl (oldv : a t) : a t option =
+    if Tbl.mem tbl @@ abstract oldv then
+      let (Abstract newv) = Tbl.find tbl (abstract oldv) in
+      match equal oldv newv with Eq -> Some newv | NotEq -> None
+    else None
+end
+
+let simplify ~partial ~context (Abstract t) =
```

Comment:
Can you add a bit of documentation here? Especially the logics to decide whether to use default condition values depending  on `partial` and which key is in `context`. It always take me a while to remember why it's working like this :-)

Reply:
I flipped the bool to `simplify ~full` at its slighly more intuitive, and added a comment.